### PR TITLE
Process unresolved Google Play service events (fix #110)

### LINKED ISSUE #110: About "GooglePlayService Dialog cancel / dismiss handling"
**Branch**: master
**Device**: Virtual device Pixel 2 XL API 29 without Play Store

### Configuration
Default, but askForGooglePlayServices is true
```
new LocationConfiguration.Builder()
                .askForPermission(new PermissionConfiguration.Builder().rationaleMessage("Gimme the permission!").build())
                .useGooglePlayServices(new GooglePlayServicesConfiguration.Builder()
                        .askForGooglePlayServices(true)
                        .build())
                .useDefaultProviders(new DefaultProviderConfiguration.Builder().gpsMessage("Would you mind to turn GPS on?").build())
                .build();
```

### Execution steps
Start "sample in activity", wait for result

### Result
Dialog is open:

After click OK button, empty activity screen is showed, `DispatcherLocationProvider#continueWithDefaultProviders()` is not called.

Empty screen, cannot get location.

### Expected behavior
After click OK button, `onCancelListener` in `DispatcherLocationProvider#resolveGooglePlayServices(int)` is called and `DispatcherLocationProvider#continueWithDefaultProviders()` invoked.

We got location from `DefaultLocationProvider`.

### Logs
```
I/LocationManager: We got permission!
W/GooglePlayServicesUtil: Google Play Store is missing.
I/DispatcherLocationProvider: GooglePlayServices is NOT available on device.
I/DispatcherLocationProvider: Asking user to handle GooglePlayServices error...
E/GoogleApiAvailability: Google Play services is invalid. Cannot recover.
```

### Conclusions

Maybe this is not library error, but com.google.android.gms implementation issue.

For this error (no Play Service) `GoogleApiAvailability#getErrorResolutionIntent(Context, int, String)` return null in `GoogleApiAvailability#getErrorDialog(Activity, int, int, OnCancelListener)`, so when
**DialogInterface.BUTTON_POSITIVE** is pressed, the **OnClickListener** listener do nothing.

We can call `Dialog#setOnDismissListener()` that catch all close events, but in listener we cannot define what event is happened (issue resolved or not?).

### Proposal

Don't call `GoogleApiAvailability#getErrorDialog()` and implement own dialog, with additional logic in OK button click listener. I looked dialog creation method  (`GoogleApiAvailability#zaa()`), all called methods in code is public.

### PR DESCRIPTION
**Pull request by victor-denisenko**
Fix #110

### CODE REVIEW

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java [hunk 1]
```diff
@@ -165,6 +165,31 @@ public void onCancel(DialogInterface dialog) {
...
+
+            /* Or
+            Intent intent = GoogleApiAvailability.getInstance().getErrorResolutionIntent(getActivity(), gpServicesAvailability, "d");
+
+            if (intent == null)
+            */
+            switch (gpServicesAvailability) {
+                case ConnectionResult.SERVICE_MISSING:
+                case ConnectionResult.SERVICE_VERSION_UPDATE_REQUIRED:
+                case ConnectionResult.SERVICE_DISABLED:
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java:177]: Let's not do exclusive, but add onDismissListener for the Result type that needs it.
That way is more intentional, and if tomorrow they introduce new types we wouldn't be adding dismiss listener without intentionally. 

```
// With a nice comment that explains this specific case why we need onDismissListener in addition to onCancelListener 
if(gpServicesAvailability == ConnectionResult.SERVICE_INVALID) {
    gpServicesDialog.setOnDismissListener(...)
}
```

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java [hunk 2]
```diff
@@ -295,6 +299,80 @@ public void resolveGooglePlayServicesShouldContinueWithDefaultWhenResolveDialogI
         verify(dispatcherLocationProvider).continueWithDefaultProviders();
     }
 
+    @Test
+    public void resolveGooglePlayServicesShouldContinueWithDefaultWhenErrorCannotBeResolved() {
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java:303]: Why is the following not enough?

```
int unresolvableError = ConnectionResult.SERVICE_INVALID;

when(dispatcherLocationSource.getGoogleApiErrorDialog(
      eq(activity), eq(unresolvableError),  eq(RequestCode.GOOGLE_PLAY_SERVICES), any(OnCancelListener.class))
).thenReturn(dialog);

dispatcherLocationProvider.resolveGooglePlayServices(unresolvableError);
verify(dialog).show();

dialog.dismiss(); // Simulate dismiss dialog (error cannot be resolved)

verify(dispatcherLocationProvider).continueWithDefaultProviders();
```

What are you trying to test by catching listeners?

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java [hunk 3]
```diff
@@ -295,6 +299,80 @@ public void resolveGooglePlayServicesShouldContinueWithDefaultWhenResolveDialogI
...
+
+        verify(dialog).show();
+
+        dialog.dismiss(); // Simulate dismiss dialog (error cannot be resolved)
+
+        verify(dispatcherLocationProvider).continueWithDefaultProviders();
+    }
+
+    @Test
+    public void resolveGooglePlayServicesShouldContinueWithDefaultWhenWhenResolveDialogIsCancelled() {
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java:342]: Same as above

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java (hunk 2 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java:303]: To check that `continueWithDefaultProviders()` is called when error is unresolvable and dialog is dismissed (pressed OK) but not cancelled.
`continueWithDefaultProviders()` called in OnDismissListener, so to simulate real listener call, I catch and store it. 
The `dialog` is `@Mock` object and `dialog.dismiss()` does nothing.

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java (hunk 3 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProviderTest.java:342]: To check that `continueWithDefaultProviders()` is called when dialog is cancelled.
`continueWithDefaultProviders()` called in OnCancelListener, so to simulate real listener call, I catch and store it. 
The `dialog` is `@Mock` object and `dialog.dismiss()` does nothing.

**Comment by yayaa** [comment]
Thank you for the changes :)

**Pull request merged**